src/types/quota.rs

```diff
@@ -0,0 +1,233 @@
+///
+/// Used by [Session::set_quota]
+pub struct QuotaResourceLimit<'a> {
+    /// The resource type
+    pub name: QuotaResourceName<'a>,
+    /// The amount for that resource
+    pub amount: u64,
+}
+
+impl Display for QuotaResourceLimit<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} {}", self.name, self.amount)
+    }
+}
+
+/// From [Resources](https://datatracker.ietf.org/doc/html/rfc2087#section-3)
+///
+/// Used by [QuotaLimit], and [QuotaResource]
+#[derive(Debug, Eq, PartialEq)]
+pub enum QuotaResourceName<'a> {
+    /// Sum of messages' RFC822.SIZE, in units of 1024 octets
+    Storage,
+    /// Number of messages
+    Message,
+    /// Any other string (for future RFCs)
+    Atom(Cow<'a, str>),
+}
+
+impl Display for QuotaResourceName<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Storage => write!(f, "STORAGE"),
+            Self::Message => write!(f, "MESSAGE"),
+            Self::Atom(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+/// From [QUOTA Response](https://datatracker.ietf.org/doc/html/rfc2087#section-5.1)
+///
+/// Due to an implementation detail this is a wrapper around [Quota] with the same contract.
+/// Access to the wrapped [Quota] can be obtained via [QuotaResponse::quota]
+///
+/// Used by [Session::get_quota] and [Session::set_quota]
+#[self_referencing]
+pub struct QuotaResponse {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) quota: Quota<'this>,
+}
+
+impl QuotaResponse {
+    /// Parse the [`Quota`] response from a response buffer.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        QuotaResponseTryBuilder {
+            data: owned,
+            quota_builder: |input| {
+                let mut quotas = Vec::new();
+
+                // There should only be one single QUOTA response
+                parse_many_into(input, &mut quotas, unsolicited, |response| match response {
+                    Response::Quota(q) => Ok(MapOrNot::Map(Quota::from_imap_proto(q))),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })?;
+
+                if quotas.is_empty() {
+                    Err(Error::Parse(ParseError::Invalid(input.to_vec())))
+                } else {
+                    Ok(quotas.remove(0))
+                }
+            },
+        }
+        .try_build()
+    }
+
+    /// The root name
+    pub fn root_name(&self) -> &str {
+        self.borrow_quota().root_name()
+    }
+
+    /// The defined resources with their usage and limits (could be empty)
+    pub fn resources(&self) -> &[QuotaResource<'_>] {
+        self.borrow_quota().resources()
+    }
+
+    /// Access to the wrapped [Quota] struct
+    pub fn quota(&self) -> &Quota<'_> {
+        self.borrow_quota()
+    }
```

Comment:
I feel like we should _either_ expose the `&Quota` _or_ accessors to the individual members (`root_name()` and `resources()`). Doing both feels a little weird. Given that this is already a `QuotaResponse`, it also feels weird to call `.quota()` on it. Thinking about it some more, how about we just `impl Deref<Target = Quota>` and then remove all three of these methods and _just_ rely on methods on `Quota` itself?

Reply:
Did you mean to add a `Deref` impl?

Reply:
@jonhoo I forgot to add a reply about that.  I TRIED to add a Deref impl, but it refused to work due to needing a lifetime specifier that couldn't resolve against the self referencing object. 

So, we have a few options
1. remove the 'quota' method but leave the other two
  - I don't like this approach as it means that the object isn't a "quota" and makes writing code that consumes a quota harder to build out.
  - We could build a trait to represent the two methods Quota has and impl that on both objects. but unsure how you feel about that.
2. fight more with the Deref impl..
  - but I have doubts that i'll work easily w/o going into unsafe {} due to the hackery in ouroboros
  - this should still allow consuming methods to work against a Quota if we can get it to work.
3. remove the two helper methods and leave JUST the quota() method.
  - This would more easily provide the access to the inner Quota object to make writing consuming methods nicer without having to add a trait

Reply:
Ah, interesting, yeah, I can see how that might get awkward. Hmm hmm hmm. I think I'm probably in favor of option 3 even though it's not very ergonomic.